### ISSUE TITLE
setup issues: use-package, helm-org-ql-actions

### ISSUE THREAD

**Issue by novoid** [toshism/org-super-links#19]
Hi,
Yes, I know that this might be totally off-topic since it's an error that `use-package` throws, not `org-super-links`. However, since I used the README of this project to start with quelpa, I think that others following these direction probably can face the very same issue. If not, I'm happy to move this to `use-package` as well.

What I did:
I set up quelpa for the first time by evaluating the following snippet:
```
(use-package quelpa
  :ensure t
)
```
Then I evaluated this:
```
(use-package org-super-links
  :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2")
  :ensure t
  :bind (("C-c s s" . sl-link)
	   ("C-c s l" . sl-store-link)
	   ("C-c s C-l" . sl-insert-link)))
```
which resulted in:
```
Debugger entered--Lisp error: (error "use-package: Unrecognized keyword: :quelpa")
  signal(error ("use-package: Unrecognized keyword: :quelpa"))
  error("use-package: %s" "Unrecognized keyword: :quelpa")
  use-package-normalize-plist(org-super-links (:quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))) nil use-package-merge-keys)
  use-package-normalize-keywords(org-super-links (:quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))))
  #f(compiled-function (name &rest args) "Declare an Emacs package by specifying a group of configuration options.\n\nFor full documentation, please see the README file that came with\nthis file.  Usage:\n\n  (use-package package-name\n     [:keyword [option]]...)\n\n:init            Code to run before PACKAGE-NAME has been loaded.\n:config          Code to run after PACKAGE-NAME has been loaded.  Note that\n                 if loading is deferred for any reason, this code does not\n                 execute until the lazy load has occurred.\n:preface         Code to be run before everything except `:disabled'; this\n                 can be used to define functions for use in `:if', or that\n                 should be seen by the byte-compiler.\n\n:mode            Form to be added to `auto-mode-alist'.\n:magic           Form to be added to `magic-mode-alist'.\n:magic-fallback  Form to be added to `magic-fallback-mode-alist'.\n:interpreter     Form to be added to `interpreter-mode-alist'.\n\n:commands        Define autoloads for commands that will be defined by the\n                 package.  This is useful if the package is being lazily\n                 loaded, and you wish to conditionally call functions in your\n                 `:init' block that are defined in the package.\n:hook            Specify hook(s) to attach this package to.\n\n:bind            Bind keys, and define autoloads for the bound commands.\n:bind*           Bind keys, and define autoloads for the bound commands,\n                 *overriding all minor mode bindings*.\n:bind-keymap     Bind a key prefix to an auto-loaded keymap defined in the\n                 package.  This is like `:bind', but for keymaps.\n:bind-keymap*    Like `:bind-keymap', but overrides all minor mode bindings\n\n:defer           Defer loading of a package -- this is implied when using\n                 `:commands', `:bind', `:bind*', `:mode', `:magic', `:hook',\n                 `:magic-fallback', or `:interpreter'.  This can be an integer,\n                 to force loading after N seconds of idle time, if the package\n                 has not already been loaded.\n:after           Delay the use-package declaration until after the named modules\n                 have loaded. Once load, it will be as though the use-package\n                 declaration (without `:after') had been seen at that moment.\n:demand          Prevent the automatic deferred loading introduced by constructs\n                 such as `:bind' (see `:defer' for the complete list).\n\n:if EXPR         Initialize and load only if EXPR evaluates to a non-nil value.\n:disabled        The package is ignored completely if this keyword is present.\n:defines         Declare certain variables to silence the byte-compiler.\n:functions       Declare certain functions to silence the byte-compiler.\n:load-path       Add to the `load-path' before attempting to load the package.\n:diminish        Support for diminish.el (if installed).\n:delight         Support for delight.el (if installed).\n:custom          Call `customize-set-variable' with each variable definition.\n:custom-face     Call `customize-set-faces' with each face definition.\n:ensure          Loads the package using package.el if necessary.\n:pin             Pin the package to an archive." #<bytecode 0x16da3b9>)(org-super-links :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link)))
  macroexpand((use-package org-super-links :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))) nil)
  macroexp-macroexpand((use-package org-super-links :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))) nil)
  macroexp--expand-all((use-package org-super-links :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))))
  macroexpand-all((use-package org-super-links :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))))
  eval-sexp-add-defvars((use-package org-super-links :quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "0.2") :ensure t :bind (("C-c s s" . sl-link) ("C-c s l" . sl-store-link) ("C-c s C-l" . sl-insert-link))))
  elisp--eval-last-sexp(nil)
  eval-last-sexp(nil)
  funcall-interactively(eval-last-sexp nil)
  call-interactively(eval-last-sexp nil nil)
  command-execute(eval-last-sexp)
```

I also evaluated `quelpa-self-upgrade` which did not change the situation. It's quelpa version `20200618.5` which is listed as installed in my `list-packages`. So the `use-package` for quelpa must have worked fine.
The issue is the same on Emacs 26.3 on Xubuntu 20.04 as well as 27.0.50 on Debian stable.

Can you give me advice here? Thanks!

**Comment by toshism** [comment]
You will also need [quelpa-use-package](https://github.com/quelpa/quelpa-use-package). I should make a note of that in the readme i guess. Thanks.

Oh, and I would also recommend using the develop branch at this point. (instead of "0.2"). I'll create a new release this week that includes quite a few changes to the configuration options.

This is my preferred setup https://github.com/toshism/org-super-links/tree/readme-format-tests#my-personal-setup-and-configuration

**Comment by novoid** [comment]
Perfect.

Installation via `use-package` went fine now.

`sl-store-link` and `sl-insert-link` are working now.

I'll allow myself for adding further issues related to the setup so that you get one (more or less clean) go-through from my side to mitigate potential issues for peers. Please correct me if you want to have separate issues or when I'm stressing your time budget too much.

`sl-link` and `sl-quick-insert-*` cause following error:
`sl-search-function: Symbol’s value as variable is void: helm-org-ql-actions`
It's a non-existent variable at my side.
Maybe you've got another idea what's happening there. I was using helm-org for various queries before. 

Your changelog lists "fix void-variable helm-org-ql-actions bug". Could this be related?

**Comment by toshism** [comment]
yes, related to #16 and #17 . I think it's most likely due to the way i'm trying to allow a dependency on _either_ helm-org-ql or helm-rifle. I'm considering just making helm-org-ql a hard dependency.

**Comment by novoid** [comment]
> What happens if you call `helm-org-ql` before calling `sl-link` to ensure it's working and loaded?

Now, I'm confused. As you can see on https://github.com/novoid/dot-emacs/blob/master/config.org I am using `helm-org` and `org-ql`. However, I don't have `helm-org-ql` and I never used it so far.
Am I still missing a package? Is `helm-org-ql` available when I first load `org-ql` and then `helm-org`?

What's the value of `helm-org-ql-actions` on your side?

(Related to the dependency: my personal opinion would be to make one of them default active and document how to switch to the other one.)

**Comment by fkgruber** [comment]
It is still not working for me. Even after installing quelpa-use-package and using the recommended config I'm getting:
```
Debugger entered--Lisp error: (wrong-type-argument arrayp nil)
  tar--describe-as-link(nil)
  tar--check-descriptor(nil)
  tar-get-file-descriptor("PaxHeader/PaxHeader-pkg.el")
  package-tar-file-info()
  quelpa-get-package-desc("/Users/fred/.emacs.d/quelpa/packages/org-super-links-20200611.950.tar")
  quelpa-package-install((org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop"))
  apply(quelpa-package-install (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop") nil)
  quelpa((org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop"))
  apply(quelpa (org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop"))
  (if (and quelpa-use-package-inhibit-loading-quelpa (package-installed-p (quote org-super-links))) nil (apply (quote quelpa) (quote ((org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop")))))
  (progn (if (and quelpa-use-package-inhibit-loading-quelpa (package-installed-p (quote org-super-links))) nil (apply (quote quelpa) (quote ((org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop"))))) (defvar use-package--warning22 (function (lambda (keyword err) (let ((msg (format "%s/%s: %s" ... keyword ...))) (display-warning (quote use-package) msg :error))))) (condition-case err (progn (if (fboundp (quote sl-link)) nil (autoload (function sl-link) "org-super-links" nil t)) (if (fboundp (quote sl-store-link)) nil (autoload (function sl-store-link) "org-super-links" nil t)) (if (fboundp (quote sl-insert-link)) nil (autoload (function sl-insert-link) "org-super-links" nil t)) (if (fboundp (quote sl-quick-insert-drawer-link)) nil (autoload (function sl-quick-insert-drawer-link) "org-super-links" nil t)) (if (fboundp (quote sl-quick-insert-inline-link)) nil (autoload (function sl-quick-insert-inline-link) "org-super-links" nil t)) (eval-after-load (quote org-super-links) (function (lambda nil (condition-case err (progn ... t) (... ...))))) (progn (let* ((name "C-c s s") (key (if ... name ...)) (kdesc (cons ... ...)) (binding (lookup-key ... key))) (let ((entry ...) (details ...)) (if entry (setcdr entry details) (add-to-list ... ...))) (define-key (or nil global-map) key (function sl-link))) (let* ((name "C-c s l") (key (if ... name ...)) (kdesc (cons ... ...)) (binding (lookup-key ... key))) (let ((entry ...) (details ...)) (if entry (setcdr entry details) (add-to-list ... ...))) (define-key (or nil global-map) key (function sl-store-link))) (let* ((name "C-c s C-l") (key (if ... name ...)) (kdesc (cons ... ...)) (binding (lookup-key ... key))) (let ((entry ...) (details ...)) (if entry (setcdr entry details) (add-to-list ... ...))) (define-key (or nil global-map) key (function sl-insert-link))) (let* ((name "C-c s d") (key (if ... name ...)) (kdesc (cons ... ...)) (binding (lookup-key ... key))) (let ((entry ...) (details ...)) (if entry (setcdr entry details) (add-to-list ... ...))) (define-key (or nil global-map) key (function sl-quick-insert-drawer-link))) (let* ((name "C-c s i") (key (if ... name ...)) (kdesc (cons ... ...)) (binding (lookup-key ... key))) (let ((entry ...) (details ...)) (if entry (setcdr entry details) (add-to-list ... ...))) (define-key (or nil global-map) key (function sl-quick-insert-inline-link))))) ((debug error) (funcall use-package--warning22 :catch err))))
  eval((progn (if (and quelpa-use-package-inhibit-loading-quelpa (package-installed-p (quote org-super-links))) nil (apply (quote quelpa) (quote ((org-super-links :repo "toshism/org-super-links" :fetcher github :commit "develop"))))) (defvar use-package--warning22 (function (lambda (keyword err) (let ((msg ...)) (display-warning (quote use-package) msg :error))))) (condition-case err (progn (if (fboundp (quote sl-link)) nil (autoload (function sl-link) "org-super-links" nil t)) (if (fboundp (quote sl-store-link)) nil (autoload (function sl-store-link) "org-super-links" nil t)) (if (fboundp (quote sl-insert-link)) nil (autoload (function sl-insert-link) "org-super-links" nil t)) (if (fboundp (quote sl-quick-insert-drawer-link)) nil (autoload (function sl-quick-insert-drawer-link) "org-super-links" nil t)) (if (fboundp (quote sl-quick-insert-inline-link)) nil (autoload (function sl-quick-insert-inline-link) "org-super-links" nil t)) (eval-after-load (quote org-super-links) (function (lambda nil (condition-case err ... ...)))) (progn (let* ((name "C-c s s") (key ...) (kdesc ...) (binding ...)) (let (... ...) (if entry ... ...)) (define-key (or nil global-map) key (function sl-link))) (let* ((name "C-c s l") (key ...) (kdesc ...) (binding ...)) (let (... ...) (if entry ... ...)) (define-key (or nil global-map) key (function sl-store-link))) (let* ((name "C-c s C-l") (key ...) (kdesc ...) (binding ...)) (let (... ...) (if entry ... ...)) (define-key (or nil global-map) key (function sl-insert-link))) (let* ((name "C-c s d") (key ...) (kdesc ...) (binding ...)) (let (... ...) (if entry ... ...)) (define-key (or nil global-map) key (function sl-quick-insert-drawer-link))) (let* ((name "C-c s i") (key ...) (kdesc ...) (binding ...)) (let (... ...) (if entry ... ...)) (define-key (or nil global-map) key (function sl-quick-insert-inline-link))))) ((debug error) (funcall use-package--warning22 :catch err)))) nil)
  elisp--eval-last-sexp(nil)
  eval-last-sexp(nil)
  funcall-interactively(eval-last-sexp nil)
  call-interactively(eval-last-sexp nil nil)
  command-execute(eval-last-sexp)

```

> You will also need [quelpa-use-package](https://github.com/quelpa/quelpa-use-package). I should make a note of that in the readme i guess. Thanks.
> 
> Oh, and I would also recommend using the develop branch at this point. (instead of "0.2"). I'll create a new release this week that includes quite a few changes to the configuration options (I also plan to add this to melpa after that).
> 
> This is my preferred setup https://github.com/toshism/org-super-links/tree/readme-format-tests#my-personal-setup-and-configuration

**Comment by toshism** [comment]
> **Update 2020-07-03: #16 and #17 do reflect my issues. So I'll stop adding comments to this issue until they seem to be solved somehow.**
> 
> > What happens if you call `helm-org-ql` before calling `sl-link` to ensure it's working and loaded?
> 
> Now, I'm confused. As you can see on https://github.com/novoid/dot-emacs/blob/master/config.org I am using `helm-org` and `org-ql`. However, I don't have `helm-org-ql` (it's not available when I do `M-x`) and I never used it so far to find out it's missing.
> 
> Am I still missing a package?
> 
> Is `helm-org-ql` available when I first load `org-ql` and then `helm-org`?
> 

So I _think_ this is actually related to `org-ql` and not `org-super-links` directly if `M-x` `helm-org-ql` is not working for you. I believe the way it works is if helm-org is installed when you install org-ql then helm-org-ql gets enabled? Maybe try reinstalling org-ql after helm-org-ql. 

> What's the value of `helm-org-ql-actions` on your side?
> 

```
(("Show heading in source buffer" . helm-org-ql-show-marker)
 ("Show heading in indirect buffer" . helm-org-ql-show-marker-indirect)
 ("Super Link" . sl-insert-link-org-ql-action))
```

**Comment by toshism** [comment]
> tar-get-file-descriptor("PaxHeader/PaxHeader-pkg.el")

@fkgruber I don't think your issue is related to this one. I would guess it's actually something with quelpa and/or tar maybe?

regardless I'll try to get this on melpa soon so there won't be an issue dealing with quelpa

**Comment by novoid** [comment]
> So I _think_ this is actually related to `org-ql` and not `org-super-links` directly if `M-x` `helm-org-ql` is not working for you. I believe the way it works is if helm-org is installed when you install org-ql then helm-org-ql gets enabled? Maybe try reinstalling org-ql after helm-org-ql.

You're a hero.

I executed `M-x list-packages`, manually uninstalled `org-ql`, manually installed `org-ql` (using the very same version) once again and then `C-c s s` worked so that I was able to set a link + backlink.

@fkgruber  as much as I understand from Elisp and Emacs, your issue also does not seem to relate to `org-super-link` since its compressed file could not be expanded. Therefore, I second the opinion of @toshism. I hope you don't mind when I close this issue for now. I wish you all the best for your issue with quelpa/tar.
